lightbeam/send.py

```diff
@@ -26,6 +41,31 @@ def send(self):
             self.logger.info("finished processing endpoint {0}!".format(endpoint))
             self.logger.info("  (final status counts: {0}) ".format(self.lightbeam.status_counts))
             self.lightbeam.log_status_reasons()
+        
+        ### Create structured output results_file if necessary
+        if self.lightbeam.results_file:
+            self.end_timestamp = datetime.datetime.now()
+            self.metadata.update({"completed_at": self.end_timestamp.isoformat(timespec='microseconds')})
+            self.metadata.update({"runtime_sec": (self.end_timestamp - self.start_timestamp).total_seconds()})
+            self.metadata.update({"total_records_processed": sum(item['records_processed'] for item in self.metadata["resources"].values())})
+            self.metadata.update({"total_records_skipped": sum(item['records_skipped'] for item in self.metadata["resources"].values())})
+            self.metadata.update({"total_records_failed": sum(item['records_failed'] for item in self.metadata["resources"].values())})
+            # total up counts by message and status
+            for resource in self.metadata["resources"].keys():
```

Comment:
Consider using `dict.items()` to circumvent some of this key-nesting.
```
for resource, resource_metadata in self.metadata["resources"].items():
    for status, status_metadata in resource_metadata.get("failed_statuses", {}).items():

        total_num_errs = 0
        for message, message_metadata in status_metadata.items():
            for file, file_metadata in message_metadata["files"].items():

                num_errs = len(file_metadata["line_numbers"])
                file_metadata.update({
                    "count": num_errs,
                    "line_numbers": ",".join(map(str, file_metadata["line_numbers"])),
                })
                total_num_errs += num_errs

        status_metadata.update({"count": total_num_errs})
```
Note: I have _not_ tested this rewrite. Make sure it all still aligns before merging.